toolchain/parse/handle_import_and_package.cpp

```diff
@@ -0,0 +1,174 @@
+                      "Expected identifier after `{0}`.", Lex::TokenKind);
+    context.emitter().Emit(*context.position(), ExpectedIdentifierAfterKeyword,
+                           context.tokens().GetKind(state.token));
+    ExitOnParseError(context, state, directive);
+    return;
+  }
+
+  bool library_parsed = false;
+  if (auto library_token = context.ConsumeIf(Lex::TokenKind::Library)) {
+    auto library_start = context.tree().size();
+
+    if (!context.ConsumeAndAddLeafNodeIf(Lex::TokenKind::StringLiteral,
+                                         NodeKind::Literal)) {
+      CARBON_DIAGNOSTIC(
+          ExpectedLibraryName, Error,
+          "Expected a string literal to specify the library name.");
+      context.emitter().Emit(*context.position(), ExpectedLibraryName);
+      ExitOnParseError(context, state, directive);
+      return;
+    }
+
+    context.AddNode(NodeKind::Library, *library_token, library_start,
+                    /*has_error=*/false);
+    library_parsed = true;
+  }
+
+  auto next_kind = context.tokens().GetKind(*(context.position()));
+  if (!library_parsed && next_kind == Lex::TokenKind::StringLiteral) {
+    // If we come acroess a string literal and we didn't parse `library
+    // "..."` yet, then most probably the user forgot to add `library`
+    // before the library name.
+    CARBON_DIAGNOSTIC(MissingLibraryKeyword, Error,
+                      "Missing `library` keyword.");
+    context.emitter().Emit(*context.position(), MissingLibraryKeyword);
+    ExitOnParseError(context, state, directive);
+    return;
+  }
+
+  if (expect_api_or_impl) {
+    switch (next_kind) {
+      case Lex::TokenKind::Api: {
+        context.AddLeafNode(NodeKind::PackageApi, context.Consume());
+        break;
+      }
+      case Lex::TokenKind::Impl: {
+        context.AddLeafNode(NodeKind::PackageImpl, context.Consume());
+        break;
+      }
+      default: {
+        CARBON_DIAGNOSTIC(ExpectedApiOrImpl, Error,
+                          "Expected `api` or `impl`.");
+        context.emitter().Emit(*context.position(), ExpectedApiOrImpl);
+        ExitOnParseError(context, state, directive);
+        return;
+      }
+    }
+  }
+
+  if (!context.PositionIs(Lex::TokenKind::Semi)) {
+    context.EmitExpectedDeclarationSemi(context.tokens().GetKind(state.token));
+    ExitOnParseError(context, state, directive);
+    return;
+  }
+
+  context.AddNode(directive, context.Consume(), state.subtree_start,
+                  state.has_error);
+}
+
+auto HandleImport(Context& context) -> void {
+  auto state = context.PopState();
+
+  auto intro_token = context.Consume();
+  context.AddLeafNode(NodeKind::ImportIntroducer, intro_token);
+
+  switch (context.packaging_state()) {
+    case Context::PackagingState::StartOfFile:
+      // `package` is no longer allowed, but `import` may repeat.
+      context.set_packaging_state(Context::PackagingState::InImports,
+                                  intro_token);
```

Comment:
I wonder if we should do this in the `AfterNonPackagingDeclaration` state too, so that we don't get repeated `ImportTooLate` errors if there are multiple imports in a row.

Reply:
Hm, no, never mind, that'll make later "first non-comment line" notes point to the wrong place.

Reply:
How about this `AfterNonPackagingDeclarationImportsWarned` setup?